rllib/agents/trainer.py

```diff
@@ -914,19 +952,32 @@ def step_attempt(self) -> ResultDict:
                 with concurrent.futures.ThreadPoolExecutor() as executor:
                     train_future = executor.submit(
```

Comment:
I notice this is still not fully parallelized, like our eval() call will overlap with at most 1 train_exec_impl() right?
I imagine the logic may be cleaner if we always have train_exec_impl() run in the main thread, and simply throw self.evaluate() into a ThreadPoolExecutor() if we want ""evaluation_parallel_to_training"".

Something like:

```
while True:
    evaluate_this_iter = ...
    if evaluate_this_iter:
        if evaluation_parallel_to_training:
            self.thread_executor.submit(self.evaluate)
        else:
            self.evaluate()
    self.train_exec_impl()
```

Something like this. Just an idea.
This is definitely out of scope for this PR, but would love to see what you think or what I missed.

Reply:
I think this is how it used to work, but the problem is that the `eval` step needs to check, whether the `train` step (which is running in a thread) is done, not the other way around. So the logic is:
- start train step in a thread
- do eval (10 timesteps as per rollout_frag_len setting)
- check if train is done
-- no? -> do another 10 steps
-- yes? -> return eval results immediately, such that the next call to `train` doesn't have to be blocked by a still running eval step

Note that we probably should not use the phases "complete decoupling" as this is misleading: We still need some form of coupling as we have to synch weights before each train+eval step from local worker to all evaluation workers. It's merely the duration of the eval step that we more closely align with the train duration in this PR such that it feels like no one has to wait for the other one anymore.

I'll add some example settings to the docs as well and change the title of this PR to clarify this. I hope this explains the logic.

Reply:
done

Reply:
again totally out of scope for this PR, just discussion at this point. I am curious:

the assumption here seems to be that eval step is much faster than train step. so we let eval workers check whether train is done. kind of feel like that's not necessarily true actually. also why do another 10 steps ... won't that make different eval steps run for different amount of steps, and introduce variance to the eval result?

I think evaluation_parallel_to_training=False is a simple case. eval simply blocks train, no controversial there.
but when evaluation_parallel_to_training=True, we gotta make sure eval runs in a separate thread so the main training thread never gets blocked?
something like:

```
while True:
    train()

    if eval_is_currently_running:
        # do nothing.
        continue
    else:
        if is_there_new_eval_result:
            # report freshly available eval results.
            eval_result = ...
        else:
            eval_result = None

        if should_we_kick_off_a_new_eval:
            sync_weights
            kick_off_a_new_eval_in_thread_pool.
```

again, not nit picking, but saying, and it feels like train in main and eval on the side is easier for users' mental model.

Reply:
It's also more likely to be CORRECT... :smirk: